File: annotation_ir.py
import ast
from typing import List
from dimension_ir import DimIR
from span import SourceSpan


class AnnotationHeadIR:
    def __init__(self, root: str, attrs: list[str], scope_id: int, span: SourceSpan):
        self.root = root        # "torch"
        self.attrs = attrs      # ["Tensor"]
        self.scope_id = scope_id
        self.span = span


class AnnotationIR:
    def __init__(self, head: AnnotationHeadIR, args: List[DimIR, DimIR]):
        self.head=head
        self.args=args

File: rust/src/ir/nodes/annotation_ir.rs
use crate::ir::{expr_ir::ExprIR, span_ir::SourceSpan};

#[derive(Debug, Clone)]
pub struct AnnotationIR {
    pub head: AnnotationHeadIR,
    pub args: Vec<ExprIR>,
}

#[derive(Debug, Clone)]
pub struct AnnotationHeadIR {
    pub root: String,
    pub attrs: Vec<String>,
    pub scope_id: i64,
    pub span: Option<SourceSpan>,
}

impl AnnotationIR {
    pub fn new(head: AnnotationHeadIR, args: Vec<ExprIR>) -> Self {
        Self {
            head,
            args,
        }
    }
}